torch/distributed/distributed_c10d.py

```diff
@@ -2082,6 +2082,11 @@ def all_gather(tensor_list, tensor, group=None, async_op=False):
         [tensor([1.+1.j, 2.+2.j]), tensor([3.+3.j, 4.+4.j])] # Rank 1
 
     """
+    # `tensor_list` would be understood as "tensor or list."
```

Comment:
doc should be updated? Also, I'm not sure how much of a fan I am of an argument possibly having multiple types, and changing the behavior based on the type - it seems that it can be surprising / confusing to users. But, if we think this is the best way to consolidate the API, then I am open to it.

Reply:
Thanks for the suggestion. I updated the document. Can you please take a look? Thank you!

Also filling in a bit of context:

- In recent discussion with @ptrblck @crcrpar regarding use of the `base` feature in Megatron, it was identified that unifying the two APIs was preferred to productizing `_all_gather_base`. Also per the initial idea to unify them as mentioned by @zhaojuanmao when `_all_gather_base` was first implemented https://github.com/pytorch/pytorch/pull/56315#pullrequestreview-639152513

- I also discussed with @ptrblck and @crcrpar regarding ways to unifying the two APIs. 
Method 1: have `all_gather` accept a single tensor
Method 2: have `all_gather` accept a list, the list containing a single tensor
It appears to me that Method 1 is cleaner in semantics -- when user passes in a single tensor, it is explicit that they ask for merging the input tensors into a (larger) output tensor. Also, since this is something user passes in rather than a return of the function, there is no need for user to figure out what the return is (a list or a tensor).

- Regarding why a single function name:
I want to give user an experience that they only need to think about what output format they want, and not having to additionally think about which function to call (especially when there are already several all_gather_* APIs there). In retrospective, it may also seem that the list implementation "hijacked" the `all_gather` name, whereas the single tensor implementation can be more natural, smoother in application use flow maybe, and more performant. Hence my thought of giving the tensor output the "first-class" name of `all_gather`.

I am happy to hear about comments on the above thoughts.